### ISSUE TITLE
Test fails on imx6q board with arm with opencl 1.1 EP

### ISSUE THREAD

**Issue by Mezzano** [BVLC/caffe#4620]
Hello,

We are trying to run caffe opencl on an armv7l on a imx6q board which supports opencl 1.1 EP.

We have configured caffe to use Greentea and CLBlast.

When running OpenCLKernelCompileTest tests, 1 passed and 1 failed.

Here's the head of the output:

`Setting to use device 0
Build Status = -2 ( Err = -11 )
Log: (224:0) : error : parameter '' type cannot be void
(224:0) : warning : kernel function : 'null_kernel_float' is empty
(467:0) : error : parameter '' type cannot be void
(467:0) : warning : kernel function : 'conv_layer_spatial_phony_float' is empty
(1441:0) : error : parameter '' type cannot be void
(1441:0) : warning : kernel function : 'fft_phony_float' is empty

Sources: #ifndef __OPENCL_VERSION__
`

And here the full output:
[https://gist.github.com/Mezzano/c3c3dbca6c47a50918f93c1fec3a7999](https://gist.github.com/Mezzano/c3c3dbca6c47a50918f93c1fec3a7999)

We are using a Vivante gc2000 GPU.

Any help would be appreciated.

**Comment by naibaf7** [comment]
Seems like that particular OpenCL compiler does not like our empty "phony" / "null" kernels.
I'll see if that can be fixed.

**Comment by bhack** [comment]
Seems like that this compiler doesn't like many things :smile:. https://github.com/CNugteren/CLBlast/issues/88

**Comment by naibaf7** [comment]
Yup, absolute paranoid compiler.
Let's see if we can make it work anyways without breaking any of the other compilers...
I just pushed a compability update, so @Mezzano please retest & close this issue if fixed.

**Comment by Mezzano** [comment]
Thanks Fabien.

We've restested with your corrections.

Head of output is now:

`Setting to use device 0
Build Status = -2 ( Err = -11 )
Log: (1387:0) : error : atomic function "atomic_cmpxchg" not supported`

And full output here:

[https://gist.github.com/Mezzano/c3c3dbca6c47a50918f93c1fec3a7999](https://gist.github.com/Mezzano/c3c3dbca6c47a50918f93c1fec3a7999)

**Comment by naibaf7** [comment]
ok, also no atomic support then..
i have to check if there is a OpenCL cap filter for that.

**Comment by Mezzano** [comment]
Maybe this can help: [http://suhorukov.blogspot.be/2011/12/opencl-11-atomic-operations-on-floating.html](http://suhorukov.blogspot.be/2011/12/opencl-11-atomic-operations-on-floating.html)

**Comment by naibaf7** [comment]
@Mezzano 
Yeah not really, since your OpenCL implementation does not even have `atomic_cmpxchg`.
The only solution is to replace algorithms with atomic functions with fallback algorithms that do not use atomics, or disable the atomic kernels and their layers completely.

**Comment by Mezzano** [comment]
Isn't the problem that OpenCL doesn't have atomic functions for floats but only for ints?
[https://www.khronos.org/registry/cl/sdk/1.1/docs/man/xhtml/](https://www.khronos.org/registry/cl/sdk/1.1/docs/man/xhtml/)

Can't find from where  `error : atomic function "atomic_cmpxchg" not supported` is coming, is it from the OpenCL compiler?

Anyway, thanks for helping us. We already have a trained network with caffe that we want to run on this specific GPU.

**Comment by naibaf7** [comment]
@Mezzano 
Yes, the error message comes from the OpenCL compiler. It explicitly states that the integer atomic function (which comes from the global 32 base atomics) is not even available, so no atomic operations at all. This is most likely due to this GPU architecture not having the interconnect fabric for atomic memory commits and cache coherence.

However I have pushed a potential fix for this, so please retest. It will mean however that you can't use LibDNN's atomic kernels and also the "embed layer" from Caffe (source: `embed.cl`) will not work on that GPU either (so the test will fail).

Regards
Fabian

**Comment by Mezzano** [comment]
@naibaf7 
I've retested with your fix, it takes quite a while but it succeeds at the end:

iCar-G3-446:/dev/disk/mounted/sda1/opencl/caffe # ./build/test/test_all.testbin --gtest_filter=*OpenCLKernelCompileTest* 0
Setting to use device 0
Note: Google Test filter = *OpenCLKernelCompileTest*
[==========] Running 2 tests from 2 test cases.
[----------] Global test environment set-up.
[----------] 1 test from OpenCLKernelCompileTest/0, where TypeParam = float
[ RUN      ] OpenCLKernelCompileTest/0.TestCompile
Kernel bundle: activation: OK
Kernel bundle: auxiliary: OK
Kernel bundle: batch_reindex: OK
Kernel bundle: benchmark: OK
Kernel bundle: bias: OK
Kernel bundle: bnll: OK
Kernel bundle: channel: OK
Kernel bundle: concat: OK
Kernel bundle: contrastive_loss: OK
Kernel bundle: conv_layer_spatial: OK
Kernel bundle: conv_spatial_helper: OK
Kernel bundle: crop: OK
Kernel bundle: dropout: OK
Kernel bundle: eltwise: OK
Kernel bundle: elu: OK
Kernel bundle: embed: OK
Kernel bundle: fft: OK
Kernel bundle: fillbuffer: OK
Kernel bundle: im2col: OK
Kernel bundle: im2col_nd: OK
Kernel bundle: lrn: OK
Kernel bundle: lstm_unit: OK
Kernel bundle: math: OK
Kernel bundle: mergecrop: OK
Kernel bundle: pooling: OK
Kernel bundle: pooling_nd: OK
Kernel bundle: pooling_sk: OK
Kernel bundle: slice: OK
Kernel bundle: softmax_loss: OK
Kernel bundle: solvers: OK
Kernel bundle: tile: OK
[       OK ] OpenCLKernelCompileTest/0.TestCompile (882 ms)
[----------] 1 test from OpenCLKernelCompileTest/0 (884 ms total)

[----------] 1 test from OpenCLKernelCompileTest/1, where TypeParam = double
[ RUN      ] OpenCLKernelCompileTest/1.TestCompile
[       OK ] OpenCLKernelCompileTest/1.TestCompile (0 ms)
[----------] 1 test from OpenCLKernelCompileTest/1 (0 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 2 test cases ran. (885 ms total)
[  PASSED  ] 2 tests.

**Comment by Mezzano** [comment]
FYI.

We still have problems. 
All caffe tests don't pass, here's the output:

[https://gist.github.com/Mezzano/bb7384cb00f40c6272e24a0af7099bba](https://gist.github.com/Mezzano/bb7384cb00f40c6272e24a0af7099bba)

It seems that there are still some problems with function declarations.
We get a CLBlast error at the end.
After passing CLBlast tests, here's what we get from dmesg:

[https://gist.github.com/Mezzano/fb2e35ec92970cafe52b063a3b46009a](https://gist.github.com/Mezzano/fb2e35ec92970cafe52b063a3b46009a)

We're digging these paths.

**Comment by naibaf7** [comment]
@Mezzano 
The first error seems like CLBlast has some code that the OpenCL compiler of your platform can't handle.
The second error seems like the OpenCL driver of your platform terminates the kernel because it thinks the kernel runs too long (hangs).

I can't really help you on these issues, since they are not Caffe specific errors and I don't have such a platform to test myself.

**Comment by simobrazz** [comment]
Hi @Mezzano,

I'm working on the same compatibility problem, did you find any solution in the last year to run Caffe on the GC2000 GPU?

Best regards,

Simone

**Comment by TechnikEmpire** [comment]
@naibaf7 What's the minimum required openCL version for the ocl branch?

**Comment by naibaf7** [comment]
@TechnikEmpire Most layers should work on OpenCL 1.1, all layers should work on OpenCL 1.2
Please note many drivers and implementations have bugs, so your mileage may vary. I usually can find a workaround for such specific bugs, however it's sometimes hard to single out issues without access to all the different hardware & operating systems.

**Comment by ekcheng** [comment]
Hi @naibaf7, so would it be accurate to say that OpenCL 1.2 is the minimum requirement? Thanks.
